provider/core/README.md

```diff
@@ -2,91 +2,103 @@
 
 `icu_provider` is one of the [`ICU4X`] components.
 
-It defines traits and structs for transmitting data through the ICU4X locale data pipeline.
-The primary trait is [`DataProvider`]. It has one method, which transforms a [`Request`] into
-a [`Response`]:
+`icu_provider` defines traits and structs for transmitting data through the ICU4X locale
+data pipeline. The primary trait is [`ResourceProvider`]. It is parameterized by a
+[`ResourceMarker`], which contains the data type and a [`ResourceKey`]. It has one method,
+[`ResourceProvider::load_resource`], which transforms a [`DataRequest`]
+into a [`DataResponse`].
 
-```rust
-fn load_payload(&self, req: &DataRequest) -> Result<DataResponse<'data>, DataError>
-```
+- [`ResourceKey`] is a fixed identifier for the data type, such as `"plurals/cardinal@1"`.
+- [`DataRequest`] contains additional annotations to choose a specific variant of the key,
+  such as a locale.
+- [`DataResponse`] contains the data if the request was successful.
 
-A [`Request`] contains a [`ResourceKey`] (a fixed identifier such as "plurals/cardinal@1") and
-[`ResourceOptions`] (a language identifier and optional variant, e.g. "fr") being requested.
-The Response contains the data payload corresponding to the Request.
+In addition, there are three other traits which are widely implemented:
 
-A [`Response`] contains a [`DataPayload`] along with other metadata.
+- [`AnyProvider`] returns data as `dyn Any` trait objects.
+- [`BufferProvider`] returns data as `[u8]` buffers.
+- [`DynProvider`] returns structured data but is not specific to a key.
 
-The most common types required for ICU4X [`DataProvider`] are included via the prelude:
+The most common types required for this crate are included via the prelude:
 
 ```rust
 use icu_provider::prelude::*;
 ```
 
-### Concrete Implementations of Data Providers
+### Types of Data Providers
+
+All data providers can fit into one of two classes.
+
+1. Type 1: Those whose data originates as structured Rust objects
```

Comment:
issue: I'm not a huge fan of "type 1" and "type 2" naming, can we just call these any providers and buffered providers?

Reply:
https://github.com/unicode-org/icu4x/issues/1576#issuecomment-1029433864